webapi/src/main/scala/org/knora/webapi/messages/StringFormatter.scala

```diff
@@ -1946,11 +1959,14 @@ class StringFormatter private (
    * @return the same string but escaped.
    */
   def validateAndEscapeProjectIri(iri: IRI, errorFun: => Nothing): IRI = // V2 / value objects
-    if (isKnoraProjectIriStr(iri)) {
-      toSparqlEncodedString(iri, errorFun)
-    } else {
-      errorFun
-    }
+    validateAndEscapeProjectIri(iri).getOrElse(errorFun)
+
+  def validateAndEscapeProjectIri(iri: IRI): Validation[ValidationException, IRI] =
+    if (isKnoraProjectIriStr(iri))
+      Validation
+        .fromOption(toSparqlEncodedString(iri))
+        .mapError(_ => ValidationException(s"Invalid IRI: $iri"))
+    else Validation.fail(ValidationException(s"Invalid IRI: $iri"))
```

Comment:
IMHO it would be totally ok to simply return an `Option` for all the functions which have taken a `errorFun: => Nothing`. This is simply because the function throwing the exception did not provide any additional information about the cause of the error case to the `errorFun`.

However, if now the method returns a `Validation` in I would expect the reason being precise. In the example of the above function in both cases the same `ValidationException` with the exact same message is returned for the three different cases of why the `iri` could not be 1. validated, 2. escaped/encoded or 3. is not a project iri.

Also on the caller side currently we do not use the `ValidationException` in those cases. So even if we provide this information it is ignored.

The same applies to the other methods you have added with a `Validation[E,A]` return type instead of using an `Option[A]`.

Reply:
Changed to `Option`.